rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,154 @@
+# cython: language_level=3, boundscheck=False
+# distutils: language = c++
+"""Bridge between Python file openers and GDAL VSI.
+
+Based on _filepath.pyx.
+"""
+
+include "gdal.pxi"
+
+import contextlib
+import logging
+from uuid import uuid4
+
+from libc.string cimport memcpy
+
+log = logging.getLogger(__name__)
+
+gdal33_version_checked = False
+gdal33_version_met = False
+
+
+# NOTE: This has to be defined outside of gdal.pxi or other C extensions will
+# try to compile C++ only code included in this header.
+cdef extern from "cpl_vsi_virtual.h":
+    cdef cppclass VSIFileManager:
+        @staticmethod
+        void* GetHandler(const char*)
+
+
+# Prefix for all in-memory paths used by GDAL's VSI system
+# Except for errors and log messages this shouldn't really be seen by the user
+cdef str PREFIX = "/vsipyopener/"
+cdef bytes PREFIX_BYTES = PREFIX.encode("utf-8")
+
+# This is global state for the Python filesystem plugin. It currently only
+# contains path -> PyOpenerBase (or subclass) instances. This is used by
+# the plugin to determine what "files" exist on "disk".
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_OBJS = set()
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+    callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+    callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+    callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+    callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+    callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+    callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+
+    if VSIFileManager.GetHandler("") == VSIFileManager.GetHandler(PREFIX_BYTES):
+        log.debug("Installing Python opener handler plugin...")
+        return VSIInstallPluginHandler(PREFIX_BYTES, callbacks_struct)
+    else:
+        return 0
+
+
+cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    if callbacks_struct is not NULL:
+        callbacks_struct.pUserData = NULL
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+    callbacks_struct = NULL
+
+
+cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) with gil:
+    """Access files in the virtual filesystem.
+
+    This function is mandatory in the GDAL Filesystem Plugin API.
+    GDAL may call this function multiple times per filename and each
+    result must be seperately seekable.
+    """
+    if pszAccess != b"r" and pszAccess != b"rb":
```

Comment:
Is there a reason why just read-only is supported?

Reply:
@groutr in my experience we read much more data than we write, so I want to get the first 80% correct first. Maybe we add writing later. It's more complicated, which is (I think) part of the reason why GDAL's [RFC 76: OGR Python drivers](https://gdal.org/development/rfc/rfc76_ogrpythondrivers.html) is read-only.

Reply:
> It's more complicated, which is (I think) part of the reason why GDAL's [RFC 76: OGR Python drivers](https://gdal.org/development/rfc/rfc76_ogrpythondrivers.html) is read-only.

Write support wasn't in the scope of the funder of that work.

For this PR, I guess write support would only be adding the write() callback.

Reply:
Let's do the write callback in a follow-up PR.